src/solvers/refinement/string_refinement.cpp

```diff
@@ -1011,31 +1070,75 @@ bool string_refinementt::check_axioms()
     }
   }
 
-  // tells whether one of the not_contains constraint can be violated
-  bool violated_not_contains=false;
+  // Maps from indexes of violated not_contains axiom to a witness of violation
+  std::map<std::size_t, exprt> violated_not_contains;
+
   debug() << "there are " << not_contains_axioms.size()
           << " not_contains axioms" << eom;
-
-  for(auto nc_axiom : not_contains_axioms)
+  for(size_t i=0; i<not_contains_axioms.size(); i++)
   {
-    typet index_type=nc_axiom.s0().length().type();
-    exprt zero=from_integer(0, index_type);
-    exprt witness=generator.get_witness_of(nc_axiom, zero);
+    const string_not_contains_constraintt &nc_axiom=not_contains_axioms[i];
+    const exprt univ_bound_inf=nc_axiom.univ_lower_bound();
+    const exprt univ_bound_sup=nc_axiom.univ_upper_bound();
+    const exprt prem=nc_axiom.premise();
+    const exprt exists_bound_inf=nc_axiom.exists_lower_bound();
+    const exprt exists_bound_sup=nc_axiom.exists_upper_bound();
+    const string_exprt s0=nc_axiom.s0();
+    const string_exprt s1=nc_axiom.s1();
+
+    symbol_exprt univ_var=generator.fresh_univ_index(
+      "not_contains_univ_var", nc_axiom.s0().length().type());
+    exprt witness=generator.get_witness_of(nc_axiom, univ_var);
     exprt val=get(witness);
-    violated_not_contains=true;
+    string_not_contains_constraintt nc_axiom_in_model(
+      get(univ_bound_inf),
+      get(univ_bound_sup),
+      get(prem),
+      get(exists_bound_inf),
+      get(exists_bound_sup),
+      to_string_expr(get(s0)),
+      to_string_expr(get(s1)));
+
+    satcheck_no_simplifiert sat_check;
```

Comment:
I would say putting all of it in a function would be best for readability, namely:
```java
    const string_constraintt &axiom=universal_axioms[i];
    symbol_exprt univ_var=axiom.univ_var();
    exprt bound_inf=axiom.lower_bound();
    exprt bound_sup=axiom.upper_bound();
    exprt prem=axiom.premise();
    exprt body=axiom.body();

    string_constraintt axiom_in_model(
      univ_var, get(bound_inf), get(bound_sup), get(prem), get(body));

    satcheck_no_simplifiert sat_check;
    supert solver(ns, sat_check);
    solver.set_ui(ui);
    add_negation_of_constraint_to_solver(axiom_in_model, solver);

    switch(solver())
    {
    case decision_proceduret::resultt::D_SATISFIABLE:
      {
        exprt val=solver.get(axiom_in_model.univ_var());
        debug() << "string constraint can be violated for "
                << axiom_in_model.univ_var().get_identifier()
                << " = " << from_expr(ns, "", val) << eom;
        violated[i]=val;
      }
      break;
    case decision_proceduret::resultt::D_UNSATISFIABLE:
      break;
    default:
      throw "failure in checking axiom";
    }
```